fastar/update_rules.py

```diff
@@ -21,66 +20,218 @@ def take(arr, idxs):
-    outmask = onp.equal(onp.dot(a_mask.astype(int), b_mask.astype(int)), onp.shape(b_mask)[0])
-    new_mask = outmask & ~old_outmask
-    out_slices = util.mask_to_slices(new_mask)
-    for slc in out_slices:
-        a_slice = (slc[0], slice(None, None)) if len(slc) > 0 else (slice(None, None),)
-        b_slice = (slice(None, None), slc[1]) if len(slc) > 1 else (slice(None, None),)
-        assert np.all(outval[slc] == 0)
-        outval = _add_at(outval, slc, lax.dot_p.bind(a[a_slice], b[b_slice]))
-    return fa.Parray((outval, outmask))
-
-unary_functions = [lax.sin_p]
-for f in unary_functions:
-    fa.update_rules[f] = unop_update(f)
-
-binary_functions = [lax.add_p, lax.sub_p, lax.mul_p]
-for f in binary_functions:
-    fa.update_rules[f] = binop_update(f)
+    return sliceableop_update(
+        lax.dot_p, old_out, a, b, output_mask=onp.equal(
+            onp.dot(a_mask.astype(int), b_mask.astype(int)),
+            onp.shape(b_mask)[0]),
+        input_slices_from_output_slice=lambda s: (
+            (s[0], slice(None)) if len(s) > 0 else (slice(None),),
+            (slice(None), s[1]) if len(s) > 1 else (slice(None),)))
+
 
 fa.update_rules[lax.dot_p] = dot_update
 
+
+def dot_general_update(old_out, a, b, dimension_numbers):
+    a, a_mask = a
+    b, b_mask = b
+    ((a_contracting_dims, b_contracting_dims),
+     (a_batch_dims, b_batch_dims)) = dimension_numbers
+
+    contraction_ratio = np.prod([a.shape[d] for d in a_contracting_dims])
+
+    o = lax.dot_general(a_mask.astype(np.float64), b_mask.astype(np.float64),
+                        dimension_numbers=dimension_numbers)
+    output_mask = onp.equal(o, onp.full_like(o, contraction_ratio))
+
+    def input_slices_from_output_slice(output_slice):
+        def result(ndim, contracting_dims, batch_dims):
+            other_dims = [i for i in range(ndim) if i not in
+                          contracting_dims and i not in batch_dims]
+
+            return tuple(slice(None) if i in contracting_dims else
+                         (output_slice[batch_dims[list(batch_dims).index(i)]]
+                          if i in batch_dims else
+                          output_slice[other_dims[other_dims.index(i)]])
+                         for i in range(ndim))
+
+        return (result(a.ndim, a_contracting_dims, a_batch_dims),
+                result(b.ndim, b_contracting_dims, b_batch_dims))
+
+    return sliceableop_update(
+        lax.dot_general_p, old_out, a, b, output_mask=output_mask,
+        input_slices_from_output_slice=input_slices_from_output_slice,
+        dimension_numbers=dimension_numbers)
+
+
+fa.update_rules[lax.dot_general_p] = dot_general_update
+
+
+def transpose_update(old_out, a, permutation):
+    a, a_mask = a
+
+    def inversely_permute_slice(s):
+        r = [slice(None)] * len(s)
+        for i, d in enumerate(permutation):
+            r[d] = s[i]
+        return tuple(r)
+
+    return sliceableop_update(
```

Comment:
I think for transpose it would (probably) be faster, or roughly as fast, to simply run `transpose(a, permutation)`, rather than updating individual elements.